### ISSUE TITLE
NGINX somehow prevents rapid testing of IIAB apps/services [e.g. during IIAB installation]

### ISSUE THREAD

**Issue by holta** [iiab/iiab#2078]
With Apache it was an absolute lifeline that I could actively test all apps during IIAB installation, using URL's like...

http://10.8.0.46/gitea
http://10.8.0.46/lokole
http://10.8.0.46/nodered
http://10.8.0.46/mediawiki
(etc)

But now with NGINX I can no longer, until an hour later after reboot or such, which makes efficient testing (i.e. a rapid feedback loop) absolutely impossible :(

FYI the equivalent URL's with http://192.168.0.191/* also fail with NGINX 404's.

Any idea why this is happening?

Tangentially Related:

#2055 'AWStats URL's erroneously redirect to http://box.lan/cgi-bin/awstats.pl?config=schoolserver'
#2056 'http://box/wiki (DokuWiki) erroneously redirects to http://127.0.0.1:8090/wiki/'
#2057 'http://box/gitea displays 404 and broken images'
#2062 'Runrole regressions [overstrict checking of local_vars.yml & let's restore dashes in role names]'
#2077 'NGINX: http://box/mediawiki erroneously redirects to http://localhost/mediawiki/index.php/Main_Page'
#2079 'NGINX: http://box/maps "404 Not Found"'

**Comment by tim-moody** [comment]
so is the problem that you get 404s or that a reboot (and a wait) are required for the links to work?
did you clear your cache?

**Comment by tim-moody** [comment]
we need to document which services are nginx direct and which have shims

**Comment by holta** [comment]
> so is the problem that you get 404s or that a reboot (and a wait) are required for the links to work?

The hourlong is the problem.  The above URL's stop showing 404s and being to work "an hour or more later" &mdash; but this destroys the rapid feedback look needed for efficient/iterative testing.

**Comment by holta** [comment]
@jvonau is the solution something as simple as running `systemctl restart nginx` (or similar) near the end of each playbook?

**Comment by jvonau** [comment]
> @jvonau is the solution something as simple as running `systemctl restart nginx` (or similar) near the end of each playbook?

No, not that simple, the above 4 urls have apache as the backend (think you could add /nextcloud /wordpress /elgg to the list) without native support so the nginx 'shims' would need to be installed in the role in addition to adding systemctl restart nginx.

> we need to document which services are nginx direct and which have shims

 It's been started see https://github.com/iiab/iiab/blob/master/roles/nginx/files/README.md

**Comment by holta** [comment]
> https://github.com/iiab/iiab/tree/master/roles/nginx/files/README.md

Great beginning.

PS that README needs to be visible from a level higher (https://github.com/iiab/iiab/tree/master/roles/nginx) as with all other playbooks.

**Comment by tim-moody** [comment]
someone please create a definitive list of 'shims' so we can check them off as we migrate to nginx direct

**Comment by holta** [comment]
> someone please create a definitive list of 'shims' so we can check them off as we migrate to nginx direct

Did @georgejhunt create these?

Can someone point to an example shim?

**Comment by jvonau** [comment]
yes, shim = contains proxy_pass http://127.0.0.1:8090/

grep 8090 /etc/nginx/conf.d/*
/etc/nginx/conf.d/dokuwiki-nginx.conf:  proxy_pass http://127.0.0.1:8090//wiki;
/etc/nginx/conf.d/elgg-nginx.conf:  proxy_pass http://127.0.0.1:8090/elgg;
/etc/nginx/conf.d/lokole-nginx.conf:  proxy_pass http://127.0.0.1:8090/lokole;
/etc/nginx/conf.d/mediawiki-nginx.conf:  proxy_pass http://127.0.0.1:8090/mediawiki;
/etc/nginx/conf.d/mediawiki-nginx.conf:    proxy_pass http://127.0.0.1:8090;
/etc/nginx/conf.d/moodle-nginx.conf:    proxy_pass http://127.0.0.1:8090;
/etc/nginx/conf.d/moodle-nginx.conf:    proxy_pass http://127.0.0.1:8090;
/etc/nginx/conf.d/nextcloud-nginx.conf:  proxy_pass http://127.0.0.1:8090/nextcloud;
/etc/nginx/conf.d/nextcloud-nginx.conf:    proxy_pass http://127.0.0.1:8090;
/etc/nginx/conf.d/nodered-nginx.conf:  proxy_pass http://127.0.0.1:8090/nodered;
/etc/nginx/conf.d/wordpress-nginx.conf:  proxy_pass http://127.0.0.1:8090/wordpress;
/etc/nginx/conf.d/wordpress-nginx.conf:    proxy_pass http://127.0.0.1:8090;

The late addition of lokole is missing from the readme.

**Comment by holta** [comment]
@tim-moody's PR #2080 ("change services to nginx and simplify php") seems to have been one attempt to solve this apparently ?

**Comment by tim-moody** [comment]
working on another. very painful.

**Comment by holta** [comment]
Fixing this means refining the (7 or so?) affected playbooks.  Context from @jvonau:

> I tested the install with `nginx_enabled: False` to test the underlying web apps worked under apache before installing the nginx shims and native support then switching to nginx with `nginx_enabled: True` and using ./runrole... allowing back to back comparisons for evaluation and further refinement. The files only_nginx.yml contains what would be in the role for nginx and disable.yml contains the code to enable the apache version of the site. The disable.yml uses a2ensite (should of used *_enable there for each app) which requires the role to [have] completed first.

**Comment by holta** [comment]
@jvonau do you recall which 3-or-so others are also affected, beyond this list of 4?

- http://box/gitea
- http://box/lokole
- http://box/nodered
- http://box/mediawiki

**Comment by holta** [comment]
[Clarified above](https://github.com/iiab/iiab/issues/2078#issuecomment-568212029) by @jvonau:

> add /nextcloud /wordpress /elgg to the list

**Comment by tim-moody** [comment]
mediawiki and wordpress are converted to native nginx in #2086

**Comment by tim-moody** [comment]
> No, not that simple, the above 4 urls have apache as the backend (think you could add /nextcloud /wordpress /elgg to the list) without native support so the nginx 'shims' would need to be installed in the role in addition to adding systemctl restart nginx.


This is an example from mediawiki:

```
- name: Remove mediawiki-nginx.conf if not mediawiki_enabled (debuntu)
  file:
    path: "{{ nginx_config_dir }}/mediawiki-nginx.conf"
    state: absent
  when: not mediawiki_enabled and is_debuntu

- name: Restart nginx service to enable/disable http://box{{ mediawiki_url }}
  systemd:
    name: nginx
    daemon_reload: yes
    state: restarted
```

**Comment by holta** [comment]
If NGINX was fully moved from [Stage 9-local-addons](https://github.com/iiab/iiab/wiki/IIAB-Contributors-Guide#ansible) to [Stage 3-base-server](https://github.com/iiab/iiab/wiki/IIAB-Contributors-Guide#ansible) (with Apache) then this would become possible.

* some shims might have to be moved around, into certain roles?

**Comment by jvonau** [comment]
#2143 moves apache and niginx to stage 4 but other wise this is more or less fixed.

**Comment by holta** [comment]
This NGINX regression is largely fixed now, thanks to @jvonau's PR #2131 and my follow-on PR #2139.

As the 2 remaining/outstanding issues here are contained and generally less serious:

1) Fixing / mitigating #2145 ("http://box/elgg only partially loads when accessed from IIAB's WAN side [then fixes itself later on during IIAB install]") is probably still very possible, but not as high a priority.

2) Whereas #2148 ("http://box/munin is installed in iiab-install's Stage 8 but shows "403 Forbidden" until various points in Stage 9")) generally fixes itself within a minute or two, so acceptable as is, despite its very misleading "403 Forbidden" browser output.
